src/ifd.rs

```diff
@@ -0,0 +1,175 @@
+use std::io::Cursor;
+
+use byteorder::{LittleEndian, ReadBytesExt};
+use tiff::decoder::ifd::{Directory, Value};
+use tiff::tags::Tag;
+
+use crate::cursor::ObjectStoreCursor;
+
+/// A collection of all the IFD
+// TODO: maybe separate out the primary/first image IFD out of the vec, as that one should have
+// geospatial metadata?
+pub(crate) struct ImageFileDirectories {
+    /// There's always at least one IFD in a TIFF. We store this separately
+    image_ifds: Vec<ImageIFD>,
+
+    // Is it guaranteed that if masks exist that there will be one per image IFD? Or could there be
+    // different numbers of image ifds and mask ifds?
+    mask_ifds: Option<Vec<MaskIFD>>,
+}
+
+impl ImageFileDirectories {
+    pub(crate) async fn open(cursor: &mut ObjectStoreCursor, ifd_offset: usize) -> Self {
+        let mut next_ifd_offset = Some(ifd_offset);
+
+        let mut image_ifds = vec![];
+        let mut mask_ifds = vec![];
+        while let Some(offset) = next_ifd_offset {
+            let ifd = ImageFileDirectory::read(cursor, offset).await;
+            next_ifd_offset = ifd.next_ifd_offset();
+            match ifd {
+                ImageFileDirectory::Image(image_ifd) => image_ifds.push(image_ifd),
+                ImageFileDirectory::Mask(mask_ifd) => mask_ifds.push(mask_ifd),
+            }
+        }
+
+        Self {
+            image_ifds,
+            // TODO: if empty, return None
+            mask_ifds: Some(mask_ifds),
+        }
+    }
+}
+
+fn value_as_usize(value: &Value) -> usize {
+    match value {
+        Value::Byte(v) => *v as usize,
+        Value::Short(v) => *v as usize,
+        Value::Signed(v) => *v as usize,
+        Value::SignedBig(v) => *v as usize,
+        Value::Unsigned(v) => *v as usize,
+        Value::UnsignedBig(v) => *v as usize,
+        _ => panic!("Not an integer"),
+    }
+}
+
+struct OptionalTags {
+    directory: Directory,
+}
+
+impl OptionalTags {
+    /// Check if the IFD contains a full resolution image (not an overview)
+    fn is_full_resolution(&self) -> bool {
+        if let Some(val) = self.directory.get(&Tag::NewSubfileType) {
+            // if self.NewSubfileType.value[0] == 0:
+            todo!()
+        } else {
+            true
+        }
+        // self.directory.contains_key(T)
+    }
+}
+
+/// An ImageFileDirectory representing Image content
+// TODO: required tags should be stored as rust-native types, not Value
+struct ImageIFD {
+    // Required tags
+    bits_per_sample: Value,
+    compression: Value,
```

Comment:
Agree that we should use rust-native types here where possible instead of `Value`. Ideally these should use (or be in) https://docs.rs/tiff/0.9.1/tiff/tags/index.html#enums, but ok to keep them like this for now.